locallib.php

```diff
@@ -621,6 +624,248 @@ private function process_action_manual_allocation() {
+        if (empty($placesleft)) {
+            // If we have no choice to assign, we return -1 to signal the algorithm that we cannot assign the user.
+            return -1;
+        }
+
+        // We now have to decide which choice id will be returned as the one the user will be assigned to.
+        // In case of "equal distribution" we have to fake the amount of available places first.
+        if ($distributionalgorithm == ACTION_DISTRIBUTE_UNALLOCATED_EQUALLY) {
+            $userstodistributecount = count($cachedundistributedusers);
+            $freeplacescount = array_reduce($placesleft, fn($a, $b) => $a + $b);
+
+            $freeplacesoverhang = $freeplacescount - $userstodistributecount;
+
+            if ($freeplacesoverhang > 0) {
+                // Only if there are more free places than users to distribute, we want to distribute "equally".
+                // Choices with more places left should be targeted first when reducing places left.
+                arsort($placesleft);
+                $i = 0;
+                $choicesmaxed = [];
+                // We now lower each count of available places in each choice for every additional place that we have altogether
+                // than users to still distribute.
+                while ($freeplacesoverhang > 0 && count(array_unique($choicesmaxed)) < count($placesleft)) {
+                    // Second condition means that we will stop if we failed trying to reduce *every* choice.
+                    $nextchoiceid = array_keys($placesleft)[$i];
+                    if ($placesleft[$nextchoiceid] > 0) {
+                        // If we can still lower it, we do it.
+                        $placesleft[$nextchoiceid] = $placesleft[$nextchoiceid] - 1;
+                        $freeplacesoverhang--;
+                    } else {
+                        // If we cannot lower the places left anymore for this choice, we track that and will try to lower the
+                        // available places for the next one instead.
+                        $choicesmaxed[] = $nextchoiceid;
+                    }
+                    $i++;
+                    // We are iterating over all the choices constantly and try to reduce the available places.
+                    $i = $i % count($placesleft);
+                }
+                // We recalculated the left places for each choice, so we have to remove the choices which are now maxed out.
+                $placesleft = array_filter($placesleft, fn($numberoffreeplaces) => $numberoffreeplaces != 0);
+            }
+        }
+
+        // From here on it's just the algorithm 'distribute by filling up'.
+        $possiblechoices = $placesleft;
+
+        $choicessortedwithgroupscount = [];
+        $choicessorted = [];
+        foreach (array_keys($possiblechoices) as $choiceid) {
+            $choicessortedwithgroupscount[count($this->get_choice_groups($choiceid))][] = $choiceid;
```

Comment:
It should be checked first whether there exists an active group restriction as the groups are not deleted directly when the restriction is disabled.

Reply:
I'm not 100% sure, if I understand your point. But are you aware of lines 735-737:
```
            if (empty($choice->usegroups)) {
                // If we have a group without group restrictions it will always be available.
                continue;
            }
```

Reply:
Yes, but these lines do not change that these choices are still considered here. The problem is that whenever you have a choice with group restrictions e.g. for two groups and you change the restriction by disabling it, those two group restrictions are still in the database. So when `get_choice_groups` is called, this would return those two groups although this is actually a choice which does not have an active group restriction any longer. Consequently, those choices are handled differently for the allocation as they are sorted by their group count. Therefore, there should be a check first at this point if `empty($choice->usegroups)`

Reply:
I again dove into the code and you are absolutely right. In my mind (and unfortunately only there ;-)) `get_choice_groups` returns an empty array in case the choice is not active, but that's obviously not the case. I think I also did not handle the case that a choice is disabled, right? I will have to sort this out as well in this algorithm?

Reply:
Just realized, I did respect disabled choices, so everything should be fine. I fixed what you asked for. Very good catch! Thanks!